File: 2116. Check if a Parentheses String Can Be Valid.java
// Medium
// Math
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/check-if-a-parentheses-string-can-be-valid/

class Solution {
    public boolean canBeValid(String s, String locked) {
        int n = s.length();
        int mx = 0, mn = 0;
        for (int i = 0; i < n; i++) {
            if (locked.charAt(i) == '1') {
                int diff;
                if (s.charAt(i) == '(') {
                    diff = 1;
                } else {
                    diff = -1;
                }
                mx += diff;
                mn = Math.max(mn + diff, (i + 1) % 2);
            } else {
                mx++;
                mn = Math.max(mn - 1, (i + 1) % 2);
            }
            if (mx < mn) return false;
        }
        return mn == 0;
    }
}

File: 2001-3000/2116. Check if a Parentheses String Can Be Valid.py
# Time:O(n), Space:O(1)

class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        n = len(s)
        if n % 2 != 0:
            return False

        balance = 0
        flexible = 0

        for i in range(n):
            if locked[i] == "0":
                flexible += 1
            elif s[i] == "(":
                balance += 1
            else:
                if balance > 0:
                    balance -= 1
                elif flexible > 0:
                    flexible -= 1
                else:
                    return False

        balance = 0
        flexible = 0

        for i in range(n - 1, -1, -1):
            if locked[i] == "0":
                flexible += 1
            elif s[i] == ")":
                balance += 1
            else:
                if balance > 0:
                    balance -= 1
                elif flexible > 0:
                    flexible -= 1
                else:
                    return False

        return True
